e2e/playground.py

```diff
@@ -0,0 +1,2076 @@
+            loc=loc,
+        )
+        self.loc_container = loc_container
+
+
+class _InputWithLabel(_InputWithContainer):
+    def __init__(
+        self,
+        page: Page,
+        *,
+        id: str,
+        loc: InitLocator,
+        loc_container: InitLocator = "div.shiny-input-container",
+        loc_label: InitLocator = "label",
+    ):
+        super().__init__(
+            page,
+            id=id,
+            loc_container=loc_container,
+            loc=loc,
+        )
+
+        if isinstance(loc_label, str):
+            loc_label = self.loc_container.locator(loc_label)
+        self.loc_label = loc_label
+
+    def value_label(self, *, timeout: Timeout = None) -> typing.Union[str, None]:
+        return self.loc_label.text_content(timeout=timeout)
+
+    def expect_label_to_have_text(self, value: str, *, timeout: Timeout = None):
+        playwright_expect(self.loc_label).to_have_text(value, timeout=timeout)
+
+
+# Class definitions
+# * Fields
+#   * Try to mirror playwright as much as possible.
+#   * There are no properties, only methods; This allows for timeout values to be passed through and for complex methods.
+#     * Locators will stay as properties
+#   * Don't sub-class. For now, use `_` separatation and use `loc` or `value` as a prefix
+# * Approach
+#   * Use locators / playwright_expect as much as possible
+#     * It should not be necessary to use `assert` directly.
+#     * MUST wait for `Locator`s to do their job
+#   * Provide `value` methods as a convenience
+
+
+class _WidthLoc:
+    def value_width(
+        self: _InputBaseP,
+        *,
+        timeout: Timeout = None,
+    ) -> typing.Union[str, None]:
+        return str_attr(self.loc, "width", timeout=timeout)
+
+    def expect_width_to_have_value(
+        self: _InputBaseP,
+        value: typing.Union[AttrValue, None],
+        *,
+        timeout: Timeout = None,
+    ):
+        expect_attr(self.loc, "width", value=value, timeout=timeout)
+
+
+class _WidthContainer:
+    def value_width(
+        self: _InputWithContainerP, *, timeout: Timeout = None
+    ) -> typing.Union[str, None]:
+        return str_attr(self.loc_container, "width", timeout=timeout)
+
+    def expect_width_to_have_value(
+        self: _InputWithContainerP,
+        value: typing.Union[AttrValue, None],
+        *,
+        timeout: Timeout = None,
+    ):
+        expect_attr(self.loc_container, "width", value=value, timeout=timeout)
+
+
+class InputNumeric(
+    _WidthLoc,
+    _InputWithLabel,
+):
+    # id: str,
+    # label: TagChildArg,
+    # value: float,
+    # *,
+    # min: Optional[float] = None,
+    # max: Optional[float] = None,
+    # step: Optional[float] = None,
+    # width: Optional[str] = None,
+    def __init__(self, page: Page, id: str):
+        super().__init__(
+            page,
+            id=id,
+            loc=f"input#{id}[type=number].shiny-bound-input",
+        )
+
+    def set(self, value: float, *, timeout: Timeout = None):
```

Comment:
Allow value to be None (the `input_numeric` function should do that too)

Reply:
ee4ae48 and 331266d